Source/WebCore/loader/ResourceMonitorPersistence.cpp

```diff
@@ -0,0 +1,194 @@
+        closeDatabase();
+    };
+
+    if (!m_sqliteDB->open(path, SQLiteDatabase::OpenMode::ReadWriteCreate, SQLiteDatabase::OpenOptions::CanSuspendWhileLocked))
+        return reportErrorAndCloseDatabase("open database"_s);
+
+    if (!m_sqliteDB->tableExists(recordTableName)) {
+        if (!m_sqliteDB->executeCommandSlow(createTableSQL(recordTableName, recordTableBody)))
+            return reportErrorAndCloseDatabase("create `record` table"_s);
+
+        RESOURCEMONITOR_RELEASE_LOG("openDatabase: Table %s created", recordTableName.characters());
+    }
+
+    if (!m_sqliteDB->indexExists(hostIndexName)) {
+
+        if (!m_sqliteDB->executeCommandSlow(createIndexSQL(hostIndexName, hostIndexBody)))
+            return reportErrorAndCloseDatabase("create `host` index on `record` table"_s);
+
+        RESOURCEMONITOR_RELEASE_LOG("openDatabase: Index %s created", hostIndexName.characters());
+    }
+
+    if (!m_sqliteDB->indexExists(accessIndexName)) {
+
+        if (!m_sqliteDB->executeCommandSlow(createIndexSQL(accessIndexName, accessIndexBody)))
+            return reportErrorAndCloseDatabase("create `access` index on `record` table"_s);
+
+        RESOURCEMONITOR_RELEASE_LOG("openDatabase: Index %s created", accessIndexName.characters());
+    }
+
+    auto insertStatement = m_sqliteDB->prepareHeapStatementSlow(insertSQL(recordTableName, "host,access"_s, "?, ?"_s));
+    if (!insertStatement)
+        return reportErrorAndCloseDatabase("prepare insert statement"_s);
+
+    m_insertSQLStatement = insertStatement.value().moveToUniquePtr();
+
+    m_sqliteDB->turnOnIncrementalAutoVacuum();
+}
+
+void ResourceMonitorPersistence::deleteExpiredRecords(ApproximateTime now, Seconds duration)
+{
+    auto expirationTime = (now - duration).secondsSinceEpoch().value();
+
+    auto deleteStatement = m_sqliteDB->prepareStatementSlow(deleteSQL(recordTableName, "WHERE access < ?"_s));
+    if (!deleteStatement || deleteStatement->bindDouble(1, expirationTime) != SQLITE_OK || !deleteStatement->executeCommand())
+        return reportSQLError("deleteExpiredRecords"_s, "delete expired records"_s);
+
+    RESOURCEMONITOR_RELEASE_LOG("deleteExpiredRecords: Deleted expired records");
+}
+
+Vector<ResourceMonitorPersistence::Record> ResourceMonitorPersistence::importRecords()
+{
+    auto selectStatement = m_sqliteDB->prepareStatementSlow(selectSQL(recordTableName, "host,access"_s, "ORDER BY access"_s));
+    if (!selectStatement) {
+        reportSQLError("importRecords"_s, "fetch records of host and access"_s);
+        return { };
+    }
+
+    Vector<Record> records;
+    while (selectStatement->step() == SQLITE_ROW) {
+        const auto& host = selectStatement->columnText(0);
+        const auto& access = selectStatement->columnDouble(1);
+
+        records.append({ host, ApproximateTime::fromRawSeconds(access) });
+    }
+
+    return records;
+}
+
+void ResourceMonitorPersistence::recordAccess(const String& host, ApproximateTime time)
```

Comment:
`ApproximateTime` is not serializable since the epoch resets to zero every time you reboot your device (i.e. `mach_approximate_time()` will be close to 0 every time you reboot). The same goes with `MonotonicTime`.

Only wall clock timestamps can be persisted. You can use `approximateWallTime()` to convert the `ApproximateTime` to a `WallTime`.